### ISSUE TITLE
cfg: hot configuration reloading

### ISSUE THREAD

**Issue by vdmit11** [tempesta-tech/tempesta#51]
We need to change some configuration options while Tempesta FW is running.

For example, we want to add a new back-end server or tune DoS protection settings, but in the same time we don't want to stop serving client connections even for a short time.

Therefore:
1. The reloading feature should be added to the Configuration Framework. It should listen a "reload" event via sysctl, and then re-parse the configuration file and push the data to Tempesta moduels without stopping the whole system. 
2. Perhaps we would not like to support the feature for all configuration options. Often it is better to have immutable configuration data in runtime and don't care about synchronization. Hence, the feature must be optional, and only some of configuration options will support it.
3. All existing code that requires the feature should be updated. At least, the following three modules should be updated:
   - 'sock_backend'  - need to change list of back-end servers without stopping the sytem
   - 'sched_http' - need to change load-balancing rules
   - 'req_conn_limit' - need to be able to tune limits on a running system

**Comment by krizhanovsky** [comment]
With #67 in mind the issue is just to make the configuration update persistent, such that the configuration file is updated as well. The particular implementation must be discussed. Probably tdbfs (#66) can be used for this, i.e. the file is stored in TDB and can be updated and viewed by a system administrator as plain ASCII file while it's actually a set of key-value pairs in TDB table and ```rest/admin``` (#67) just writes appropriate values to the table on configuration reloading.

**Comment by vdmit11** [comment]
We already discussed that, but just for the note:

> make the configuration update persistent, such that the configuration file is updated as well. 

It seems the configuration file is an one-way mechanism: it can be parsed easily, but you can't update it in-place if something changed.

The update is hard because when you parse the file, you are loosing a lot of information (positions, formatting, comments, nested structures, etc). If you want to update it in-place, you have to preserve this information (store a parsed tree in the memory all the time). You also have to update thins in an artificial way: you don't want to break formatting or delete comments.

This is a complex and buggy way that requires a lot of effort.
It is much easier to achieve persistence by storing changes in a well-structured database.


> TDB and can be updated and viewed by a system administrator as plain ASCII file while it's actually a set of key-value pairs in TDB table and rest/admin (#67) just writes appropriate values to the table on configuration reloading.

Well, a TDB table can be viewed as a file easily, but things go complex when you allow the user to edit this file:
1. Suppose a user changed some values in the file. How do you know what actually was changed? You need to know which TDB entries should be updated. Hence, you have to create a `diff` of the old/new version of the file, and map certain sections of the file to TDB entries, and so on.
2. What if a TDB entry is changed while the file is being edited? Keeping in mind the `1.`, you have to update the "old" version of the file and TDB mappings. That surely will conflict with a "new" file version saved by the user.
3. You can't preserve formatting, comments and use `include` directives in the configuration file. This is a big disadvantage for a system administrator.

**Comment by vdmit11** [comment]
As a solution problems pointed in the previous comment, I suggest to NOT update the file at all.
Configuration files are designed for editing by humans, so let it be.

Despite of that, we need some configuration CLI for testing. System adminstrators will find that useful too (for scripting/automation).

I suggest to create a command-line REST client, something like `/usr/bin/tempesta_cfg` that will be able to read/write/list configuration options.
For example, you can use it in a bash script as follows:
```
tempesta_cfg list
tempesta_cfg get listen
tempesta_cfg set listen :8081
```
Very handy for scripting.
The same interface can be used for testing and system administration.


If you think carefully about it, you don't even need a configuration file parser.
The configuration file may just look like:
```
listen :80
backend :8080
cache on
```
That can be easily interpreted by a bash script like:
```
tempesta_cfg set listen :80
tempesta_cfg set backend :8080
tempesta_cfg set cache on
```
The configuration parser now is trivial.
As a bonus, you have the same interface for: configuration file, testing, system administration.

**Comment by krizhanovsky** [comment]
Should be done in user-space application via FastCGI interface (#77) and using some simple template engine like [Mustache](http://mustache.github.io/)

**Comment by vdmit11** [comment]
Alexander, as far as I understand, FastCGI imples that the Tempsta FW is a front-end communicating with a back-end application via the FastCGI interface. The Tempsta FW sends requests, and the application sends back responses. 

I think this is a little bit complicated because this is one-direction channel: the backend userspace application can't push something to the Tempesta FW. That menas we have to maintain separate mechanisms for loading configuration files to Tempesta FW and saving them back to the user space.

I suggest a simple (but a little bit radical) solution. I think we should remove state from Tempesta FW and move configuration handling to userspace as much as possible. I'll try to describe my vision of how should it be implemented.

**1. REST as a general approach**

You can think of the configuration parsing process as a process of restoring the application state. The state is transferred from the file to the application. REST is designed specially for that task, so I think we can use it.

Currently Tempesta FW modules declare their own sections and directives in the configuration file. As we already discussed, this approach has a bunch of problems: one-way direction, hot reloading, etc.

I suggest to get rid of the "file" concept at all. Modules should export their resources via a RESTful API, and the configuration subsytem should act as a client accessing those resources.

Here is the diagram:
```
+-------------------------+      +-------------------------------------------------------------------------------------+
| A configuration client  |      |  The main Tempesta FW kernel module                                                 |
+-------------------------+      +-------------------------------------------------------------------------------------+
| Various clients may be  |      |                                                                                     |
| implemented, such as:   |      |  +--------------------------------------+       +---------------------------------+ |
|  * file parser          |      |  | tempesta_fw/rest.c                   |       | tempesta_fw/backend.c           | |
|  * command-line tool    |      |  +--------------------------------------+       +---------------------------------+ |
|  * Web UI               |      |  |                                      |       |                                 | |
|                      +------------->tfw_rest_dispatch(method, uri, data);|   +---->tfw_backends_dispatch()         | |
| They can be separate    |      |  |    +                                 |   |   |   +                             | |
| kernel modules.         |      |  |    |    +--------------------------+ |   |   |   +->GET+----> list_backends(); | |
+-------------------------+      |  |    |    | Dispatch List            | |   |   |   |                             | |
                                 |  |  search +-----------------+--------+ |   |   |   +->POST+---> add_backend();   | |
                                 |  |    |    |"/cache/"        |callback| |   |   |   |                             | |
                                 |  |    |    +--------------------------+ |   |   |   +->DELETE+-> del_backend();   | |
                                 |  |    +---->"/backends/"     |callback+-----+   |                                 | |
                                 |  |         +--------------------------+ |       +---------------------------------+ |
                                 |  |         |"/filter/"       |callback| |                                           |
                                 |  |         +--------------------------+ |                                           |
+-------------------------+      |  |         |"/frang/req_rate"|callback| |                                           |
|An external module       |      |  |     add +-----------------+--------+ |                                           |
|(e.g. classifier/frang.c)|      |  |    +---->                          | |                                           |
+-------------------------+      |  |    |    +--------------------------+ |                                           |
|                         |      |  |    +                                 |                                           |
| module_init();+--------------------->tfw_rest_register(uri, callback);   |                                           |
|                         |      |  |                                      |                                           |
+-------------------------+      |  +--------------------------------------+                                           |
                                 |                                                                                     |
                                 +-------------------------------------------------------------------------------------+

```

The RESTful API (`tempesta_fw/rest.c`) I described above here is just a tiny RPC-like mechanism. It has nothing to do with the HTTP (although the HTTP front-end may be implemented on top of it). 

I suggest to leave the only  dispatcher in the core kernel module. Everything else (including the currently implemented file parser) may be moved to separate kernel modules or to userspace (which is even better).

Also, this API should eliminate any need for `sysctl` and `debugfs` hooks that are currently used to export some "variables" to userspace. Now we should be able to export them as resources via REST.
The whole `debugfs.c` was created by me to do that for testing purposes. Surely, we can get rid of it and use the API for testing. Moreover, we can use the same API for testing, configuration and system administration.

**2. kernel/user interfaces and configuration management*

As I mentioned before, REST doesn't mean HTTP over TCP.

It is possible to implement multiple userspace interfaces for different purposes.
For example, it may look like:
```
                                   +                                                                   
                             User  |  Kernel                                                           
                                   |                                                                   
+----------------------------+     |   +------------------------------+                                
|web_admin.js (in a browser) |     |   |rest_if_http.c                |                                
+----------------------------+     |   +------------------------------+                                
| jQuery.ajax();             |     |   |In kernel web ser er          |                                
| <form method="post">       |     |   +---------------------+--------+                                
| etc...                    +--------> |listen()/accept()/recv()      |                                
+---------------------------++     |   |parse_http_request();         |                                
                                   |   |extract_method_uri_and_data();|                                
                                   |   |send_to_tfw();                +------+                         
+----------------------------+     |   +------------------------------+      |                         
|/usr/bin/tfw_parsefile      |     |                                         |                         
+----------------------------+     |   +------------------------------+      |                         
|open();                     |     |   | rest_if_netlink.c            |      |                         
|while(parse_cfg_entry())    |     |   +------------------------------+      +-> +--------------------+
|     sendmsg();             +--+  |   |Custom binary protocol        |          |tempesta_fw/rest.c  |
|close();                    |  |  |   |o er Netlink socket.          |   +----> +--------------------+
+----------------------------+  |  |   +---+--------------------------+   |      |tfw_rest_dispatch();|
                                +----> |recvmsg();                    |   |  +-> +--------------------+
+----------------------------+     |   |unpack_args();                |   |  |                         
|/usr/bin/tfw_ctl            |  +----> |send_to_tfw();                +---+  |                         
+----------------------------+  |  |   +------------------------------+      |                         
|parse_command_line_args();  |  |  |                                         |                         
|sendmsg();                  +--+  |   +------------------------------+      |                         
+----------------------------+     |   | rest_if_fs.c                 |      |                         
                                   |   +--------------+---------------+      |                         
                                   |   | A simple HTTP-like text proto|      |                         
+----------------------------+     |   | col o er sysctl/debugfs/etc. |      |                         
| t/test_filter.sh           |     |   +------------------------------+      |                         
+----------------------------+     |   |handle_read();                |      |                         
|echo "POST /filter/rules/"  +-------> |parse_request();              |      |                         
| ^ /proc/sys/tempesta_fw/ctl|     |   |send_to_tfw();                +------+                         
+-+--------------------------+     |   +------------------------------+                                
                                   +                                                                   
```

All the different things listed above are just exampes that illustrate approach. In reality (at least for now) we don't need so many interfaces and clients.

I think we can leave the single `netlink` interface, and also collect all configuration-related tools into the single user-space package, like this:
```
                                                                                    +                                                 
                                                                              User  |  Kernel                                         
                                                                                    |                                                 
                                           +-----------------------------------+    |                                                 
                                           | tfw_manager package               |    |                                                 
                                           +-----------------------------------+    |                                                 
                  +--------------+         |                                   |    |                                                 
                  |Web Browser   |         | +---------------+                 |    |                                                 
                  +-------------++         | |Web SerVer     |                 |    |                                                 
               +-->jQuery.ajax()|----------> +---------------+     +---------+ | netlink+------------------+    +--------------------+
               |  +----------+---|         | |admin.py       +----->libtfwmgr+------+--->rest_if_netlink.c +--> |tfw_rest_dispatch();|
               |  |admin.hml <---------------+admin.html     |     +^--^-----+ |    |   +------------------+    +--------------------+
               +--+admin.js  <---------------|admin.js       |      |  |       |    |                                                 
                  |logo.png  <---------------+logo.png       |      |  |       |    |   +------------------+                          
                  +----------+---+         | +---------------+      |  |       |    |   |Other rest_if_*   |                          
                                           |                        |  |       |    |   |modules may be    |                          
+----------------------------------+       | +---------------+      |  |       |    |   |implemented later |                          
|tempesta.sh                       |       | |/usr/bin/tfwcfg|      |  |       |    |   +------------------+                          
+----------------------------------+       | +---------------+      |  |       |    |                                                 
|tfwcfg load /etc/tfw.cfg          +--------->parse_file()   |      |  |       |    |                                                 
|tfwctl restart                    |       | |send_to_tfw(); +------+  |       |    |                                                 
+----------------------------------+---+   | +---------------+         |       |    |                                                 
                                       |   |                           |       |    |                                                 
+----------------------------------+   |   | +---------------+         |       |    |                                                 
|test_filter.sh                    |   +---> |/usr/bin/tfwctl|         |       |    |                                                 
+----------------------------------+       | +---------------+         |       |    |                                                 
|tfwctl add /filters/ IP=127.0.0.1 +-------> |parse_args();  |         |       |    |                                                 
+----------------------------------+       | |send_to_tfw(); +---------+       |    |                                                 
                                       +---> +---------------+                 |    |                                                 
+----------------------------------+   |   |                                   |    |                                                 
|/etc/cron.weekly/backup.sh        |   |   +-----------------------------------+    |                                                 
+----------------------------------+   |                                            |                                                 
|tfwctl stop                       +---+                                            +                                                 
|backup                            |                                                                                                  
|tfwctl start                      |                                                                                                  
+----------------------------------+                                                                                                  
```

This approach has several advantages:
 - Most of the code is executed in the userspace, hence the debugging and easy, and the kernel module doesn't bloat.
 - Good isolation: you can test the Configuration Manager and Tempesta FW separately.
 - The Configuration Manager may be implemented in a high-level language like Perl or Python and use a modern high-level server-side web framework like Django or RubyOnRails. 
 - It also can use a template engine like Mustache to update the configuration files when the configuration is changed via Web UI. That should be much easier now since everything is handled by the same userspace application. 

**3. Persistency**

In my opinion, we should NOT use TDB to store configuration when the Tempesta FW is stopped. 

The main problem with the approach is it makes the whole thing stateful: you can't anymore restart Tempesta FW and just load a fresh configuration into it, you always have to track what has been changed.

Another problem is you have to synchronize TDB and files all the time. The tdbfs is a solution, but it is harder to implement. Things are generally easier to implement in userspace, so I suggest to move the configuration storage to userspace completely.

Of course, that doesn't mean that you won't use TDB to store data. Tempesta FW can save some data there, but not configuration.

I think it is better to store configuration in userspace, either files or database, and load everything via the RESTful API. I think that should make things easier, especially if you want to update configuration files from Web UI.

A short illustration is:
```
                    +--------------------------+                               
                    |                          |                               
+----------------+  |   +------------------+   |                               
|tfw_cfg_manager +--+--->A template engine |   |                               
|Web serVer/UI   |      |(Mustache, etc)   |  PUT                              
+----------------+      +-----+------------+   |                               
                              |                |     +-----------------+       
                            update             +----->Tempesta FW      |       
                              |                      |kernel module    |       
                          +---v----+           +----->REST oVer netlink|       
                          |tfw.cfg |           |     +-----------------+       
                          +---^----+           |                               
                              |               POST                             
                            parse              |                               
                              |                |                               
+-----------------+     +-----+------------+   |                               
|tempesta.sh start+----->/usr/bin/tfwcfg   +---+                               
+-----------------+     +------------------+                                   
```

Here the configuration is always stored in files. A more complicated scheme with files+database may be implemented. In any case it is easier to update the file in userspace.

Clustering is also easier in userspace. We can share configuration files via a distributed FS easily, or use database replication, or we can have a dedicated Web UI machine and distribute REST queries as HTTP requests across machines running Tempesta FW.

Ok, I think it is enough to present my vision of the right way to do the configuration. 

@krizhanovsky, @keshonok, please comment.

**Comment by krizhanovsky** [comment]
This is undoable for complex configuration language to generate configuration file using templates. At least formating will be broken. Since Tempesta is introducing it's own DSL for configuration and runtime functionality, it's easier modify the whole program in context of #67 and compare ASTs of new and old program and reload (binary opcode/bytecode patch) the only different nodes.

**Comment by ikoveshnikov** [comment]
closed via 3f49adaf2ecb55a67b343896eebaeef2caadbd68

**Comment by ikoveshnikov** [comment]
closed via 3f49adaf2ecb55a67b343896eebaeef2caadbd68
